### PR TITLE
JAMES-1718 add an implementation for AccessTokenRepository

### PR DESCRIPTION
**Pull request by chibenwa**
Because it would be nice to have support for distributed authentication !

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java [hunk 1]
```diff
@@ -43,6 +44,7 @@
         new CassandraDomainListModule(),
         new CassandraRecipientRewriteTableModule(),
         new CassandraSieveRepositoryModule(),
+        new CassandraJmapModule(),
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java]: should be located into a new project data-jmap-guice or something similar.

File: server/data/data-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    private final Session session;
+    private final PreparedStatement removeStatement;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectStatement;
+    private final int durationInSeconds;
+
+    @Inject
+    public CassandraAccessTokenRepository(Session session, @Named(TOKEN_EXPIRATION_IN_MS) long durationInMilliseconds) {
+        this.session = session;
+        this.durationInSeconds = (int) (durationInMilliseconds / 1000);
```
Comment (mbaechler) [inline server/data/data-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java]: Ints.checkedCast

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java (hunk 1 above)
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java]: I don't understand this message.
Creating a new Maven project ? Why ? What will be the benefits ? Why do you consider it wrong to have your Jmap data storage definition iside Cassandra project ? In what is it different from a UsersRepository, a DomainList or other interfaces used to store James data ?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Before
+    public void setUp() {
+        accessTokenRepository = createAccessTokenRepository();
+    }
+
+    abstract protected AccessTokenRepository createAccessTokenRepository();
+
+    @Test
+    public void validTokenMustWork() throws Exception {
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: what do you mean by "must work" ?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    public static final long TTL_IN_MS = 1000;
+
+    private AccessTokenRepository accessTokenRepository;
+
+    @Before
+    public void setUp() {
+        accessTokenRepository = createAccessTokenRepository();
+    }
+
+    abstract protected AccessTokenRepository createAccessTokenRepository();
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: abstract comes after protected

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test
+    public void validTokenMustWork() throws Exception {
+        accessTokenRepository.addToken(USERNAME, TOKEN);
+        assertThat(accessTokenRepository.getUsernameFromToken(TOKEN)).isEqualTo(USERNAME);
+    }
+
+    @Test(expected=InvalidAccessToken.class)
+    public void nonStoredTokensMustBeInvalid() throws Exception {
+        accessTokenRepository.getUsernameFromToken(TOKEN);
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: you can use assertThatThrownBy instead

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    abstract protected AccessTokenRepository createAccessTokenRepository();
+
+    @Test
+    public void validTokenMustWork() throws Exception {
+        accessTokenRepository.addToken(USERNAME, TOKEN);
+        assertThat(accessTokenRepository.getUsernameFromToken(TOKEN)).isEqualTo(USERNAME);
+    }
+
+    @Test(expected=InvalidAccessToken.class)
+    public void nonStoredTokensMustBeInvalid() throws Exception {
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: nonStored ?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Test(expected=InvalidAccessToken.class)
+    public void removedTokensMustBeInvalid() throws Exception {
+        accessTokenRepository.addToken(USERNAME, TOKEN);
+        accessTokenRepository.removeToken(TOKEN);
+        accessTokenRepository.getUsernameFromToken(TOKEN);
+    }
+
+    @Test(expected = AccessTokenAlreadyStored.class)
+    public void addTokenMustThrowWhenTokenIsAlreadyStored() throws Exception {
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: are you sure it makes sense ?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        try {
+            accessTokenRepository.addToken(USERNAME, TOKEN);
+        } catch(Exception e) {
```
Comment (mbaechler) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: you don't need to write ugly things like that with assertThatThrownBy

**Comment by mbaechler** [comment]
what about async queries ? and a split between logic and access layers ?

File: server/data/data-jmap-cassandra/pom.xml [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+                    <groupId>org.mockito</groupId>
+                    <artifactId>mockito-core</artifactId>
+                    <scope>test</scope>
+                </dependency>
+                <dependency>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-api</artifactId>
+                </dependency>
+                <dependency>
+                    <groupId>org.slf4j</groupId>
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/pom.xml]: please don't

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+    private static final String TTL = "ttl";
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement removeStatement;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectStatement;
+    private final int durationInSeconds;
+
+    public CassandraAccessTokenDAO(Session session, long durationInMilliseconds) {
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java]: @Inject ?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java [hunk 11]
```diff
@@ -48,7 +49,7 @@
 
     public CassandraAccessTokenDAO(Session session, long durationInMilliseconds) {
         this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
-        this.durationInSeconds = (int) (durationInMilliseconds / 1000);
+        this.durationInSeconds = Ints.checkedCast(durationInMilliseconds / 1000);
```
Comment (mbaechler) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java]: It's no the last solution I suggested. You can use TimeUnit.millis().toSeconds

File: server/data/data-jmap/pom.xml [hunk 12]
```diff
@@ -182,6 +182,10 @@
                     <artifactId>guava</artifactId>
                 </dependency>
                 <dependency>
+                    <groupId>com.jason-goodwin</groupId>
```
Comment (mbaechler) [inline server/data/data-jmap/pom.xml]: I don't see any value using Try in this changeset

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java (hunk 10 above)
Comment (fvn-linagora) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenDAO.java]: and @VisibleforTesting ?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+import org.apache.james.jmap.api.access.exceptions.InvalidAccessToken;
+import org.junit.Before;
+import org.junit.Test;
+
+public abstract class AbstractAccessTokenRepositoryTest {
+
+    private static final AccessToken TOKEN = AccessToken.generate();
+    private static final String USERNAME = "username";
+    public static final long TTL_IN_MS = 1000;
```
Comment (fvn-linagora) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/access/AbstractAccessTokenRepositoryTest.java]: protected ?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.cassandra.access;
+
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+import javax.inject.Named;
```
Comment (fvn-linagora) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java]: unused

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.apache.james.jmap.api.access.AccessToken;
+import org.apache.james.jmap.api.access.AccessTokenRepository;
+import org.apache.james.jmap.api.access.exceptions.InvalidAccessToken;
+
+import com.datastax.driver.core.Session;
```
Comment (fvn-linagora) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java]: unused

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+import com.google.common.base.Preconditions;
+import com.jasongoodwin.monads.Try;
```
Comment (fvn-linagora) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/access/CassandraAccessTokenRepository.java]: unused

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/access/AccessTokenRepository.java [hunk 17]
```diff
@@ -19,17 +19,20 @@
 
 package org.apache.james.jmap.api.access;
 
-import org.apache.james.jmap.api.access.exceptions.AccessTokenAlreadyStored;
+import java.util.concurrent.CompletableFuture;
+
 import org.apache.james.jmap.api.access.exceptions.InvalidAccessToken;
 
+import com.jasongoodwin.monads.Try;
```
Comment (fvn-linagora) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/access/AccessTokenRepository.java]: unused

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/access/MemoryAccessTokenRepository.java [hunk 18]
```diff
@@ -28,10 +30,10 @@
 import org.apache.commons.collections4.map.PassiveExpiringMap;
 import org.apache.james.jmap.api.access.AccessToken;
 import org.apache.james.jmap.api.access.AccessTokenRepository;
-import org.apache.james.jmap.api.access.exceptions.AccessTokenAlreadyStored;
 import org.apache.james.jmap.api.access.exceptions.InvalidAccessToken;
 
 import com.google.common.base.Preconditions;
+import com.jasongoodwin.monads.Try;
```
Comment (fvn-linagora) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/memory/access/MemoryAccessTokenRepository.java]: unused

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/crypto/AccessTokenManagerImpl.java [hunk 19]
```diff
@@ -19,6 +19,9 @@
 
 package org.apache.james.jmap.crypto;
 
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/crypto/AccessTokenManagerImpl.java]: unused

**Comment by fvn-linagora** [comment]
I guess you could make use of Intellij's "Inspect Code" feature, before pushing any change for review, or at least let the tool remove unused imports automatically.

**Comment by fvn-linagora** [comment]
Is there any benefit adding another layer (repository AND DAO) between persistence impl and api/service/manager/... ?
And is "repository" not misleading, since it does not look like a DDD repository ?

**Comment by chibenwa** [comment]
Ok, sure.
The checkbox "optimize import" upon commit was unchecked... I guess it might help.

**Comment by chibenwa** [comment]
For the DAO, refer to matthieu : 

```
what about async queries ? and a split between logic and access layers ?
```

I guess it will become a defacto standard.... On some piece of code, the benefit is high (CassandraSieveRepository) on some other not that much (here) but seems like consistency in our code base is more important than avoiding adding a layer.

**Comment by chibenwa** [comment]
test this please

File: server/pom.xml [hunk 20]
```diff
@@ -878,6 +891,11 @@
                 <version>${geronimo-activation-spec.version}</version>
             </dependency>
             <dependency>
+                <groupId>com.jason-goodwin</groupId>
+                <artifactId>better-monads</artifactId>
```
Comment (mbaechler) [inline server/pom.xml]: not need ?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/access/AccessToken.java [hunk 21]
```diff
@@ -48,6 +48,10 @@ public String serialize() {
         return token.toString();
     }
 
+    public UUID getToken() {
```
Comment (mbaechler) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/access/AccessToken.java]: A token can't really have a getToken method IMO. Or it should return `this`. Maybe it's more a `serialize` or `asUUID`

Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/access/AccessToken.java]: I choose asUUID

**Comment by fvn-linagora** [comment]
I guess there is confusion regarding the term 'logic' here, that can both refer to business logic and to data  access logic at the same time. But if @mbaechler is OK with that layering ...

**Comment by mbaechler** [comment]
:+1:

**Comment by fvn-linagora** [comment]
:+1:

**Pull request closed without merging**